### ISSUE TITLE
"Intrinsic & Extrinsic Sizing" feature discussion

### ISSUE THREAD

**Issue by Fyrd** [Fyrd/caniuse#5901]
I was looking into the Intrinsic & Extrinsic Sizing feature recently and I feel like due to the way the specs and implementations have changed over the years it may have become more confusing than useful.

When it was first added in 2013 and the concept of intrinsic width was new it made sense to track support even if it was implemented unevenly just so people could tell how to accomplish some the basic parts cross-browser. 

Now however there's a history of legacy values and newly specified values and of course the values will behave differently for different properties.

As far as I can tell the BCD feature tables do a good job at handling a lot if not all of these, and looking at them (through the summary of https://caniuse.com/?search=min-content for example) I notice there's a lot of differences in support by property, suggesting that lumping them together doesn't work very well.

So for now I'm considering either removing the feature altogether in favor of the BCD tables, or perhaps repurposing it to a subset of values/scenarios but I'm having a hard time finding something that wouldn't still be messy and confusing in some way that's not already solved by the BCD tables.

I don't take removing a feature lightly though, which is why I thought I'd bring it up here in search of suggestions. @Schweinepriester & @atjn as the top caniuse co-contributors I'm particularly interested in hearing your opinions. :)

**Comment by atjn** [comment]
Hmm, I agree that trying to have complete support data on so many values in one place is making it hard to both use and maintain the table, but the way I see it, tables like this is exactly why Caniuse is so useful.

If I am a developer getting into intrinsic sizing, I will at first just want to get a quick impression of whether or not it will be feasible to use. If I only had the MDN data, getting that impression would be very cumbersome, whereas the Caniuse table instantly lets me know that I probably can make it work, since it's supported in all browsers, although there is enough edge cases that I will have to thoroughly test it, and maybe use the MDN data as I move forward with the finer details of my implementation.

Having the above use case in mind, I would probably keep the support table, but scrap most of the comments and replace them with something like:
"Support is highly fragmented depending on the specific value and property it is used on. Check if your use case is supported by making a search with the name of both the property and value that you want to use (example: [`width min-content`](https://caniuse.com/?search=width%20min-content))."

Maybe it would also make sense to downgrade all browsers to `partial` support, just to make it more clear to the user that nothing is guaranteed, and they should make use of the MDN tables to be certain.

Hope this helps, happy to discuss further :)

**Comment by Fyrd** [comment]
Cool, yeah that sounds pretty good to me. I guess that would make it similar to https://caniuse.com/es6 except that I was able to set a threshold there at which point I started marking it as "supported". It would be nice if we could determine that point for this feature as well. 

Perhaps if min-content, max-content, and fit-content() (the values specified in the [level 3 spec](https://www.w3.org/TR/css-sizing-3/)) are supported across all properties that support them? The stretch, fit-content and contain values will supposedly come in level 4 so it would make sense to exclude these even though they have some amount of support today as well.

Not a huge fan of that differentiation but who knows how long everything would stay at "partial" support otherwise.

**Comment by atjn** [comment]
I don't really use these features, so It's hard to determine what makes sense, but making the level 3 spec determine full support sounds like a good solution to me :+1:

**Comment by Fyrd** [comment]
@atjn Alright, will probably go that route then. Thanks for weighing in!

@Schweinepriester do you have thoughts on this?

**Comment by Schweinepriester** [comment]
(Yeah, sorry for not responding yet, I'll focus on doing that today)

**Comment by Schweinepriester** [comment]
So, finally: I concur with the direction!

> If I am a developer getting into intrinsic sizing, I will at first just want to get a quick impression of whether or not it will be feasible to use. If I only had the MDN data, getting that impression would be very cumbersome, whereas the Caniuse table instantly lets me know that I probably can make it work, since it's supported in all browsers, although there is enough edge cases that I will have to thoroughly test it, and maybe use the MDN data as I move forward with the finer details of my implementation.

+1

> Perhaps if min-content, max-content, and fit-content() (the values specified in the [level 3 spec](https://www.w3.org/TR/css-sizing-3/)) are supported across all properties that support them? The stretch, fit-content and contain values will supposedly come in level 4 so it would make sense to exclude these even though they have some amount of support today as well.
>
> Not a huge fan of that differentiation but who knows how long everything would stay at "partial" support otherwise.

Why don't you like that differentiation?
I think that's the perfect way to go, just like ECMAScript 2015 is basically defined by the sum of the ~features, one entry for level 3 and one for level 4, pretty clear defined whats needed for full support :)

We might even be able to leverage the [web-platform-tests](https://web-platform-tests.org/) or the like to define support.

The entries could even be (semi)automated by adding all sub-features together with each data update, combining the strengths of BCD and caniuse? B-)
In other words: 
Feature request: Automated meta-entries, support defined by the sum of sub-features :D :)
(Nudge #5872 0:-))

One minor nitpick I'm sure is already on your radar: I really think the URL <https://caniuse.com/intrinsic-width> should change, i.e. move the feature, best case redirecting existing links/embeds 0:-)

**Comment by Fyrd** [comment]
Thanks!

> Why don't you like that differentiation?

I think in this particular case what bothers me is that while `min-content` and `max-content` are in the spec and have pretty good support, the `fit-content()` function value [does not](https://caniuse.com/mdn-css_properties_width_fit-content_function) (except inside grids)... instead the `fit-content` keyword value _does_ have [good support](https://caniuse.com/mdn-css_properties_width_fit-content) but is not in the current spec. I'm sure this makes sense from the spec developers POV but I can absolutely see this being confusing to developers. 

One way around that would be to only consider `min-content` and `max-content` here, certainly wouldn't be the first case that only part of a spec is tracked on a table due to varying browser support. What are your thoughts on doing that?

Regarding automating tests and leveraging WPT, definitely something I've considered, when time allows I'll look into that some more. :)

**Comment by Schweinepriester** [comment]
I see!

> One way around that would be to only consider `min-content` and `max-content` here, certainly wouldn't be the first case that only part of a spec is tracked on a table due to varying browser support. What are your thoughts on doing that?

As long as it's transparent that's fine I think. Best case linked to what's not included.
I guess that would be easy here for level 3, simply linking to <https://caniuse.com/mdn-css_properties_width_fit-content_function>.

Maybe worded something like:
"Basic support for Intrinsic & Extrinsic Sizing, a subset of [Box Sizing Module Level 3](https://www.w3.org/TR/css-sizing-3/). Includes a, b, c; excluding x, y, [`fit-content()` function value](https://caniuse.com/mdn-css_properties_width_fit-content_function)"

**Comment by Fyrd** [comment]
Hmm, seems rather wordy, I was just going to go for something like "CSS min-content & max-content values" (bit similar to the [grab & grabbing](https://caniuse.com/css3-cursors-grab) values, though obviously a lot more complicated). I think that would be more developer-friendly since most folks don't dig into the spec to see exactly what it does and does not cover.

**Comment by Schweinepriester** [comment]
Fair enough!
